jupyterhub/tests/selenium/test_browser.py

```diff
@@ -0,0 +1,419 @@
+            # will drop parameters given to the login page, passing only the next url
+            'login',
+            [('param', 'fromlogin'), ('next', '/hub/spawn?param=value')],
+            '/hub/login?param=fromlogin&next=%2Fhub%2Fspawn%3Fparam%3Dvalue',
+            '/hub/login?next=%2Fhub%2Fspawn%3Fparam%3Dvalue',
+        ),
+        (
+            # login?param=value&anotherparam=anothervalue
+            # will drop parameters given to the login page, and use an empty next url
+            'login',
+            [('param', 'value'), ('anotherparam', 'anothervalue')],
+            '/hub/login?param=value&anotherparam=anothervalue',
+            '/hub/login?next=',
+        ),
+        (
+            # login
+            # simplest case, accessing the login URL, gives an empty next url
+            'login',
+            [],
+            '/hub/login',
+            '/hub/login?next=',
+        ),
+    ],
+)
+async def test_open_url_login(app, browser, url, params, redirected_url, form_action):
+    url = url_path_join(public_host(app), app.hub.base_url, url)
+    url_new = url_concat(url, params)
+    await in_thread(browser.get, url_new)
+    redirected_url = redirected_url.replace('{{BASE_URL}}', url_escape(app.base_url))
+    form_action = form_action.replace('{{BASE_URL}}', url_escape(app.base_url))
+    form = browser.find_element(*LoginPageLocators.FORM_LOGIN).get_attribute('action')
+
+    print("Link in form: " + form)
+    # verify title / url
+    assert browser.title == LoginPageLocators.PAGE_TITLE
+    assert form.endswith(form_action)
+    print("Current Link in form: " + browser.current_url)
+    # login in with params
+    await login(browser, user='test_user', pass_w='test_user')
+    # verify next url + params
+    user = 'test_user'
+    next_url = browser.current_url
+    if url_escape(app.base_url) in form_action:
+        print("Current Link 2: " + browser.current_url)
+        assert next_url.endswith("param=value")
+    elif "next=%2Fhub" in form_action:
+        assert next_url.endswith("spawn?param=value")
+        assert f"user/{user}/" not in next_url
+        print("Current Link 2: " + browser.current_url)
+    else:
+        if next_url.endswith(f"/user/{user}/") == False:
+            webdriver_wait(
+                browser, EC.url_to_be(ujoin(public_url(app), f"/user/{user}/"))
+            )
+        assert next_url.endswith(f"/user/{user}/")
+        print("Current Link 2: " + browser.current_url)
+
+
+"""TBD"""
+
+
+@pytest.mark.parametrize(
+    'running, next_url, location, params',
+    [
+        # default URL if next not specified, for both running and not
+        (True, '', '', None),
+        (False, '', '', None),
+        # next_url is respected
+        (False, '/hub/admin', '/hub/admin', None),
+        (False, '/user/other', '/hub/user/other', None),
+        (False, '/absolute', '/absolute', None),
+        (False, '/has?query#andhash', '/has?query#andhash', None),
```

Comment:
I love seeing these comments to help indicate what various test variations parts is about!!! Nice!